pkg/cmd/monitor/monitor.go

```diff
@@ -0,0 +1,279 @@
+	signal.Notify(shutdownHandler, shutdownSignals...)
+	go func() {
+		select {
+		case <-shutdownHandler:
+			klog.Infof("Received SIGTERM or SIGINT signal, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		case <-ctx.Done():
+			klog.Infof("Context has been cancelled, shutting down.")
+			close(shutdownHandler)
+			cancel()
+		}
+	}()
+
+	// enable profiler by default
+	defer serviceability.Profile("web").Stop()
+	serviceability.StartProfiler()
+
+	lg, err := health.GetZapLogger(health.LoglevelToZap(o.logLevel), o.logOutputs, o.enableLogRotation, o.LogRotationConfigJSON)
+	if err != nil {
+		return err
+	}
+
+	// TODO make health checks configurable by flag(s)
+	monitor, err := o.newMonitor(ctx, lg,
+		WithSingleTargetHealthCheck(
+			health.WithSerializedReadSingleTarget(),
+			health.WithGRPCReadySingleTarget(),
+			health.WithQuorumReadSingleTarget(),
+			// TODO
+			// health.WithWatchRequireLeader(),
+		),
+		WithMultiTargetHealthCheck(
+			health.WithQuorumRead(),
+			// TODO
+			// health.WithNodeICMP(),
+		),
+	)
+
+	monitor.wg.Add(1)
+	go func() {
+		defer monitor.wg.Done()
+		monitor.Schedule()
+	}()
+	monitor.wg.Wait()
+	lg.Info("health monitor is shutting down",
+		zap.String("pod", o.podName),
+		zap.Int("static-pod-version", o.staticPodVersion),
+	)
+	return nil
+}
+
+// newMonitor creates a series of health checks. Health check probes can be single or multi target.
+func (o *monitorOpts) newMonitor(ctx context.Context, lg *zap.Logger, singleTargetChecks []health.CheckFunc, multiTargetChecks []health.CheckFunc) (*Monitor, error) {
+	var healthChecks []*health.Check
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      o.clientCertFile,
+		KeyFile:       o.clientKeyFile,
+		TrustedCAFile: o.clientCACertFile,
+	}
+
+	targets := strings.Split(o.Targets, ",")
+
+	// Create single target checks one check per target
+	for _, target := range targets {
+		// one client per target to eliminate lock racing
+		client, err := newETCD3Client(tlsInfo, targets)
+		if err != nil {
+			return nil, err
+		}
+		for _, healthCheckFunc := range singleTargetChecks {
+			healthCheck := health.NewCheck(ctx, lg, client, []string{target})
+			healthCheck.Probe = healthCheckFunc
+			healthChecks = append(healthChecks, healthCheck)
+		}
+	}
+
+	client, err := newETCD3Client(tlsInfo, targets)
+	if err != nil {
+		return nil, err
+	}
+	for _, healthCheckFunc := range multiTargetChecks {
+		healthCheck := health.NewCheck(ctx, lg, client, targets)
+		healthCheck.Probe = healthCheckFunc
+		healthChecks = append(healthChecks, healthCheck)
+	}
+
+	var wg sync.WaitGroup
+	monitor := &Monitor{
+		ctx:          ctx,
+		wg:           &wg,
+		interval:     o.interval,
+		healthChecks: healthChecks,
+	}
+
+	return monitor, nil
+}
+
+func (m *Monitor) Schedule() {
+	for _, healthCheck := range m.healthChecks {
+		m.wg.Add(1)
+		go func(healthCheck *health.Check) {
+			defer m.wg.Done()
+			err := wait.PollImmediateUntil(m.interval,
```

Comment:
Because this waits for one iteration to be complete before starting another, I think we need to set all the timeouts to 1s or less.  I'm ok doing it this way for simplicity of reading, but part of the point here is high granularity.

Reply:
nice +1